services/server/src/server/session.ts

```diff
@@ -0,0 +1,90 @@
+import expressSession from "express-session";
+import createMemoryStore from "memorystore";
+import { Pool } from "pg";
+import genFunc from "connect-pg-simple";
+import config from "config";
+import session from "express-session";
+
+import logger from "../common/logger";
+import { NextFunction, Request, Response } from "express";
+
+function getSessionStore() {
+  const sessionStoreType = config.get("session.storeType");
+
+  switch (sessionStoreType) {
+    case "database": {
+      const pool = new Pool({
+        host: process.env.SOURCIFY_POSTGRES_HOST,
+        database: process.env.SOURCIFY_POSTGRES_DB,
+        user: process.env.SOURCIFY_POSTGRES_USER,
+        password: process.env.SOURCIFY_POSTGRES_PASSWORD_1 || "a",
+        port: parseInt(process.env.SOURCIFY_POSTGRES_PORT || "5432"),
+      });
+
+      // This listener is necessary otherwise the sourcify process crashes if the database is closed
+      pool.prependListener("error", () => {
+        logger.error("Database connection lost for session pool");
+      });
+
+      const PostgresqlStore = genFunc(expressSession);
+
+      return new PostgresqlStore({
+        pool: pool,
+      });
+    }
+
+    case "memory": {
+      // Initialize the memory store if 'memory' is selected
+      const MemoryStore = createMemoryStore(expressSession);
+      return new MemoryStore({
+        checkPeriod: config.get("session.maxAge"),
+      });
+    }
+
+    default:
+      // Throw an error if an unrecognized session storage type is selected
+      throw new Error(
+        `Selected session storage type '${sessionStoreType}' doesn't exist.`
+      );
+  }
+}
+
+function getSessionOptions(): session.SessionOptions {
+  if (config.get("session.secret") === "CHANGE_ME") {
+    const msg =
+      "The session secret is not set, please set it in the config file";
+    process.env.NODE_ENV === "production"
+      ? logger.error(msg)
+      : logger.warn(msg);
+  }
+  return {
+    secret: config.get("session.secret"),
+    name: "sourcify_vid",
+    rolling: true,
+    resave: false,
+    saveUninitialized: true,
+    cookie: {
+      maxAge: config.get("session.maxAge"),
+      secure: config.get("session.secure"),
+      sameSite: "lax",
+    },
+    store: getSessionStore(),
+  };
+}
+
+export default function getSessionMiddleware() {
+  const sessionMiddleware = session(getSessionOptions());
+  // We need to wrap the sessionMiddleware in a middleware to prevent it from returning all the postgresql errors in the 500 request
+  return (req: Request, res: Response, next: NextFunction) => {
+    sessionMiddleware(req, res, (error) => {
```

Comment:
I have the doubt that in this way we are going to catch also all the previous middleware that execute `next(err)`

Reply:
I feel so too, but I am not sure. Did you manage to test it in some way?

Reply:
Maybe it would be possible to add a test for this case?

Reply:
> Maybe it would be possible to add a test for this case?

That's actually a good idea! But is it possible to make the memory session crash on purpose?

Reply:
I would say with mocking it should be possible, but with the current state of the server tests it's fine like this.

What I was actually referring to is if we can make a previous middleware fail. So that we test that this middleware does not catch all errors.

Reply:
Ah okok, I'll do it manually and check what happens 👍

Reply:
Ok the previous middleware errors are not catched by this middleware 👍